### PR TITLE
[FSDP] Introduce optim_state_dict() and load_optim_state_dict()

### PR DESCRIPTION
**Pull request by fegin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86912

Current FSDP saving and loading optimizer state_dict involves several different APIs. This PR aims to consolidate the usage and make the APIs looks like FSDP.state_dict()/FSDP.load_state_dict() by introducing FSDP.optim_state_dict() and FSDP.load_optim_state_dict().

Differential Revision: [D40083733](https://our.internmc.facebook.com/intern/diff/D40083733/)

### CODE REVIEW

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -2110,7 +2191,8 @@ def set_state_dict_type(
         module: nn.Module,
         state_dict_type: StateDictType,
         state_dict_config: Optional[StateDictConfig] = None,
-    ) -> Tuple[StateDictType, StateDictConfig]:
+        optim_state_dict_config: Optional[OptimStateDictConfig] = None,
+    ) -> Tuple[StateDictType, StateDictConfig, OptimStateDictConfig]:
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2195]: unrelated to this change - does `set_state_dict_type` need to be public or private?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -295,12 +297,21 @@ class StateDictType(Enum):
...
+                   >>>     state = model.state_dict()
             3. ``_sharded_state_dict/_load_sharded_state_dict``: this pair of APIs
                return and load sharded, unflattened parameters. The ``state_dict``
                return by ``sharded_state_dict`` can be used by all other parallel
                schemes (resharding may be required).
+    ..note::
+        The StateDictType also controls how FSDP.optim_state_dict() return the
+        states of the optimizer. The sharding strategy will be consistent for
+        ``model.state_dict()` and ``FSDP.optim_state_dict()``:
+            >>> with FSDP.state_dict_type(StateDictType.SHARDED_STATE_DICT):
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:310]: do we anticipate a use case where one state_dict is sharded and other is full, for some reason?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -2136,54 +2218,82 @@ def set_state_dict_type(
...
             expected_state_dict_config_type = _state_dict_type_to_config[state_dict_type]
             if expected_state_dict_config_type != type(state_dict_config):
                 raise RuntimeError(
                     f"Expected state_dict_config of type {expected_state_dict_config_type} "
                     f"but got {type(state_dict_config)}"
                 )
+            expected_optim_state_dict_config_type = _optim_state_dict_type_to_config[state_dict_type]
+            if expected_optim_state_dict_config_type != type(optim_state_dict_config):
+                raise RuntimeError(
+                    f"Expected state_dict_config of type {expected_optim_state_dict_config_type} "
```
[the remark this answers is not in the document]
Comment (rohan-varma) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2281]: Also, ideally we should have unittests for these validation checks. Feel free to file follow up issue to do so.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -3922,6 +4046,96 @@ def _warn_optim_input(optim_input):
...
+        state_dict_settings = FSDP.state_dict_settings(model)
+        if state_dict_settings.state_dict_type == StateDictType.FULL_STATE_DICT:
+            rank0_only = cast(
+                state_dict_settings.optim_state_dict_config, FullOptimStateDictConfig
+            ).rank0_only
+            return FSDP.full_optim_state_dict(
+                model, optim, rank0_only=rank0_only, group=group
+            )
+        elif state_dict_settings.state_dict_type == StateDictType.SHARDED_STATE_DICT:
+            return FSDP.sharded_optim_state_dict(model, optim, group=group)
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:4084]: where do we pass CPU offloading info?

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 5]
```diff
@@ -3922,6 +4046,96 @@ def _warn_optim_input(optim_input):
...
+            ``model``. The sharding of the optimizer state is based on
+            ``state_dict_type``.
+        """
+        FSDP = FullyShardedDataParallel
+        state_dict_settings = FSDP.state_dict_settings(model)
+        if state_dict_settings.state_dict_type == StateDictType.FULL_STATE_DICT:
+            rank0_only = cast(
+                state_dict_settings.optim_state_dict_config, FullOptimStateDictConfig
+            ).rank0_only
+            return FSDP.full_optim_state_dict(
```
Comment (rohan-varma) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:4080]: where do we pass CPU offloading info?

File: test/distributed/fsdp/test_fsdp_optim_state.py [hunk 6]
```diff
@@ -1259,6 +1261,47 @@ def _run_on_all_optim_state_apis(
...
+                with FSDP.state_dict_type(
+                    model2,
+                    state_dict_type,
+                    optim_state_dict_config=optim_state_dict_config,
+                ):
+                    FSDP.load_optim_state_dict(fsdp_osd1, model2, optim2)
+
+                local_osd1 = optim1.state_dict()
+                local_osd2 = optim2.state_dict()
+                self._check_same_state(
```
Comment (rohan-varma) [inline test/distributed/fsdp/test_fsdp_optim_state.py:1302]: does this check CPU offloading details?

**Review by rohan-varma — COMMENTED** [review]
thanks for adding this! have some comments.

**Review by zhaojuanmao — COMMENTED** [review]
this look good to me, thanks for the clean up!!

do we also want to update docstring to use set_state_dict_type() later sometime?

**Pull request closed without merging**